### PR TITLE
Allow setting the directory where to store the screenshots

### PR DESCRIPTION
**Pull request by jricardo27**

### CODE REVIEW

File: aloe_webdriver/screenshot_failed.py [hunk 1]
```diff
@@ -45,9 +45,30 @@
 # pylint:disable=no-member
 
 
+if 'SCREENSHOT_DIR' not in globals():
```
Comment (koterpillar) [inline aloe_webdriver/screenshot_failed.py]: Why not define the constant the normal way?

File: aloe_webdriver/screenshot_failed.py [hunk 2]
```diff
@@ -45,9 +45,30 @@
...
+
+
 FORMAT = 'failed_{feature_file}_{scenario_index}_{scenario_name}{outline_index}'
 
 
+def set_save_directory(directory):
+    """Sets the directory for saving screenshots."""
+
+    if not os.path.isdir(directory):
+        os.makedirs(directory)
```
Comment (koterpillar) [inline aloe_webdriver/screenshot_failed.py]: `os.makedirs(directory, exist_ok=True)`

File: aloe_webdriver/screenshot_failed.py [hunk 3]
```diff
@@ -81,6 +102,9 @@ def take_screenshot(self):
     )
     base_name = re.sub(r'\W', '_', base_name, flags=re.UNICODE)
 
+    if SCREENSHOT_DIR:
```
Comment (koterpillar) [inline aloe_webdriver/screenshot_failed.py]: It should be possible to achieve this with `FORMAT` alone, just use `dirname` on `base_name`.

File: aloe_webdriver/tests/test_screenshot_failed.py [hunk 4]
```diff
@@ -13,22 +13,36 @@
...
+    dir_path = None
+
     def cleanup_screenshots(self):
         """Clean up any screenshots taken."""
 
         for ext in ('png', 'html'):
             for filename in iglob('failed_*.{}'.format(ext)):
                 os.unlink(filename)
 
+        # Delete temporal directory if any.
```
Comment (koterpillar) [inline aloe_webdriver/tests/test_screenshot_failed.py]: "temporary directory", not "temporal" (here and below).

File: aloe_webdriver/screenshot_failed.py [hunk 5]
```diff
@@ -45,9 +45,30 @@
...
 
 
+if 'SCREENSHOT_DIR' not in globals():
+    SCREENSHOT_DIR = None
+
+
 FORMAT = 'failed_{feature_file}_{scenario_index}_{scenario_name}{outline_index}'
 
 
+def set_save_directory(directory):
```
Comment (koterpillar) [inline aloe_webdriver/screenshot_failed.py]: It is possible to set `aloe_webdriver.screenshot_failed.FORMAT` (and `SCREENSHOT_DIR` with your changes) directly. Why the extra functions?

File: aloe_webdriver/screenshot_failed.py [hunk 6]
```diff
@@ -45,9 +45,30 @@
...
+    """Sets the directory for saving screenshots."""
+
+    if not os.path.isdir(directory):
+        os.makedirs(directory)
+
+    global SCREENSHOT_DIR  # pylint:disable=global-statement
+    SCREENSHOT_DIR = directory
+
+
+def unset_save_directory():
```
Comment (koterpillar) [inline aloe_webdriver/screenshot_failed.py]: This looks like it's only useful within tests.

File: aloe_webdriver/screenshot_failed.py [hunk 7]
```diff
@@ -45,9 +45,30 @@
...
+    """Unset the directory fr saving screenshots."""
```
Comment (koterpillar) [inline aloe_webdriver/screenshot_failed.py]: Typo.

**Review by koterpillar — CHANGES_REQUESTED** [review]
If you want, merge `FORMAT` and `SCREENSHOT_DIR` into one and point out that it can include the directory names in the documentation.

The tricks with `globals()` shouldn't be necessary even in tests, because the module is reloaded anyway. _If_ for some reason it's unavoidable, there should be a comment explaining why.

**Comment by jricardo27** [comment]
- Preferred to leave `FORMAT` and `SCREENSHOT_DIR` separate as is easier to enforce a slash at the end of the directory by using `os.path.join`. Removed the extra `if` though.

- Updated the documentation.

- Removed unnecessary functions.

- Still need to use the trick with `globals()` as the problem is precisely the module being reloaded on tests and resetting the `SCREENSHOT_DIR` value. Added a note about that.

**Comment by koterpillar** [comment]
Interesting, I doubted that `globals` was necessary. Here's what I found:

* `reload(aloe_webdriver.screenshot_failed)` is triggered by an environment variable set in tests. I suspect in your real test suite you set the directory from a steps file too, so why not add another environment variable to control the directory? E.g.

```python

File: aloe_webdriver/screenshot_failed.py [hunk 8]
```diff
@@ -26,6 +26,17 @@
 
     failed_features_account_feature_1_Log_in.png
     failed_features_account_feature_1_Log_in.html
+
+To change the directory where the screenshots are saved, override the constant
+``SCREENSHOT_DIR`` as follow:
```
Comment (koterpillar) [inline aloe_webdriver/screenshot_failed.py]: "as follows"

File: aloe_webdriver/screenshot_failed.py [hunk 9]
```diff
@@ -47,6 +58,12 @@
 
 FORMAT = 'failed_{feature_file}_{scenario_index}_{scenario_name}{outline_index}'
 
+# This constant should be redefined externally. Due to how `run_features` in
+# `aloe.testing.FeatureTest.run_features` reload modules on each run, this check
+#  prevents it to be reset.
```
Comment (koterpillar) [inline aloe_webdriver/screenshot_failed.py]: "prevents it from being reset" (but irrelevant, because the `globals` trick isn't needed)

**Review by koterpillar — CHANGES_REQUESTED** [review]
Interesting, I doubted that `globals` was necessary. Here's what I found: `reload(aloe_webdriver.screenshot_failed)` is triggered by an environment variable set in tests. I suspect in your real test suite you set the directory from a steps file too, so why not add another environment variable to control the directory? E.g.
```python
if os.environ.get('TAKE_SCREENSHOTS'):
    # Only register the screenshot steps if asked to
    import aloe_webdriver.screenshot_failed
    reload(aloe_webdriver.screenshot_failed)

    screenshot_dir = os.environ.get('SCREENSHOT_DIR')
    if screenshot_dir:
        aloe_webdriver.screenshot_failed.SCREENSHOT_DIR = screenshot_dir
```

Also, `screenshot_failed.SCREENSHOT_DIR` is redundant, as it will always be used qualified. Rename to `DIRECTORY`.

**Comment by jricardo27** [comment]
Updated.

**Comment by koterpillar** [comment]
The build failed, please fix.

**Comment by jricardo27** [comment]
Just a couple failed due to timeout when downloading a Docker image.

How can I run them again?

```
docker: error pulling image configuration: Get https://dseasb33srnrn.cloudfront.net/registry-v2/docker/registry/v2/blobs/sha256/1a/1a2c5a5ec0336adb8bfdcd9cd5961e3fc1071899546f83fa4628b58fcd29c5c7/data?Expires=1487726891&Signature=hLmBcBC9v3xCTsWnmTAHnUC9Gh9ZhXrknn4ft5f2fPD6EDyjH6LQwm~Bf13w-Js1FarxgKlvzHvFxjNUpYkWnDVHLtk2bUAbKtyuRQCSmKj03fWIJauL1agh4M527aFqNeKLYEM2nSWDk1B7rhpb7f4Jgkfgb8q4sP9JyP9w2xo_&Key-Pair-Id=APKAJECH5M7VWIS5YZ6Q: net/http: TLS handshake timeout.
```

**Review by koterpillar — APPROVED** [review]

**Comment by koterpillar** [comment]
Merged, will be released by Travis shortly. Thanks!

**Pull request merged**